test/unit/messaging/batch-requests.spec.ts

```diff
@@ -0,0 +1,239 @@
+    const batch = new BatchRequestClient(httpClient, batchUrl, commonHeaders);
+
+    const responses: HttpResponse[] = await batch.send(requests);
+
+    expect(responses.length).to.equal(1);
+    expect(stub).to.have.been.calledOnce;
+    const args: HttpRequestConfig = stub.getCall(0).args[0];
+    expect(args.headers).to.have.property('X-Custom-Header', 'value');
+  });
+
+  it('should add sub request headers to the payload', async () => {
+    const stub = sinon.stub(httpClient, 'send').resolves(
+      createMultipartResponse([responseObject]));
+    stubs.push(stub);
+    const requests: SubRequest[] = [
+      {url: 'https://example.com', body: {foo: 1}, headers: {'X-Custom-Header': 'value'}},
+      {url: 'https://example.com', body: {foo: 1}, headers: {'X-Custom-Header': 'value'}},
+    ];
+    const commonHeaders = {'X-Custom-Header': 'value'};
+    const batch = new BatchRequestClient(httpClient, batchUrl, commonHeaders);
+
+    const responses: HttpResponse[] = await batch.send(requests);
+
+    expect(responses.length).to.equal(1);
+    expect(stub).to.have.been.calledOnce;
+    const args: HttpRequestConfig = stub.getCall(0).args[0];
+    const parsedRequest = parseHttpRequest(args.data as Buffer);
+    expect(parsedRequest.multipart.length).to.equal(requests.length);
+    parsedRequest.multipart.forEach((part: {body: Buffer}) => {
+      const parsedPart: {headers: object} = parseHttpRequest(part.body.toString().trim());
+      expect(parsedPart.headers).to.have.property('X-Custom-Header', 'value');
+    });
+  });
+
+  function checkOutgoingRequest(stub: sinon.SinonStub, requests: SubRequest[]) {
+    expect(stub).to.have.been.calledOnce;
+    const args: HttpRequestConfig = stub.getCall(0).args[0];
+    expect(args.method).to.equal('POST');
+    expect(args.url).to.equal(batchUrl);
+    expect(args.headers).to.have.property(
+      'Content-Type', 'multipart/mixed; boundary=__END_OF_PART__');
+    expect(args.timeout).to.equal(10000);
+    const parsedRequest = parseHttpRequest(args.data as Buffer);
+    expect(parsedRequest.multipart.length).to.equal(requests.length);
+
+    if (requests.length === 1) {
+      // http-message-parser handles single-element batches slightly differently. Specifically, the
+      // payload contents are exposed through body instead of multipart, and the body string uses
+      // \n instead of \r\n for line breaks.
+      let expectedPartData = getParsedPartData(requests[0].body);
+      expectedPartData = expectedPartData.replace(/\r\n/g, '\n');
+      expect(parsedRequest.body.trim()).to.equal(expectedPartData);
+    } else {
+      requests.forEach((req, idx) => {
+        const part = parsedRequest.multipart[idx].body.toString().trim();
+        expect(part).to.equal(getParsedPartData(req.body));
+      });
+    }
+  }
+});
+
+function parseHttpRequest(text: string | Buffer): any {
+  const httpMessageParser = require('http-message-parser');
```

Comment:
Suggestion: I think we should minimize depending on low usage npm modules. In this case, this has ~2k weekly downloads. We have been bitten by some dependency on an npm module that was abandoned in the past.

Reply:
We only use this in a couple of test cases. I can try to replace it with something else in a future PR.